### PR TITLE
Preliminary SourceMap support

### PR DESCRIPTION
**Pull request by jsmall-nvidia**
* Only supports decoding currently
* Added support for Rtti functions map, to allow types external to Rtti impl to work

### CODE REVIEW

File: source/slang/slang-language-server.h [hunk 1]
```diff
@@ -100,7 +100,8 @@ class LanguageServer
     Slang::InlayHintOptions m_inlayHintOptions;
     bool m_quit = false;
     List<LanguageServerProtocol::WorkspaceFolder> m_workspaceFolders;
-
+    RttiTypeFuncsMap m_typeMap;
```
Comment (csyonghe) [inline source/slang/slang-language-server.h:103]: This does not need to be initialized? I wonder when do we actually register new entries into the map.

**Review by csyonghe — APPROVED** [review]

File: source/compiler-core/slang-source-map.cpp [hunk 2]
```diff
@@ -213,7 +211,7 @@ SlangResult SourceMap::decode(JSONContainer* container, const JSONSourceMap& src
                 // 5 parts
                 if (segment.getLength() > 0)
                 {
-                    /* If present, the zero - based index into the �names� list associated with this segment.This field is a base 64 VLQ relative to the previous occurrence of this field, unless this is the first occurrence of this field, in which case the whole value is represented.
+                    /* If present, the zero - based index into the “names” list associated with this segment.This field is a base 64 VLQ relative to the previous occurrence of this field, unless this is the first occurrence of this field, in which case the whole value is represented.
```
[the remark this answers is not in the document]
Comment (csyonghe) [reply source/compiler-core/slang-source-map.cpp:214]: Nevermind, this is in extended ASCII, so we should be good.

**Review by csyonghe — APPROVED** [review]

File: source/slang/slang-language-server.h (hunk 1 above)
Comment (jsmall-nvidia) [reply source/slang/slang-language-server.h:103]: I think this is initialized in init on the LanguageServer. 

I should have elaborated more on the need for the map, which is that the Rtti assorted tools only worked with types that were "known" by the RttiUtil which didn't include JSONValue. RttiUtil is defined in core. Having JSONValue as a *native* value was useful to be able to make the distinction between null and a string value which was at best somewhat contrived with a slice or worse a string. 

For the current Language server, the use of types "outside" of RttiUtil worked ok. It didn't work though with the List of JSONValue I was using inside the native representation of the SourceMap type. 

Short version - if it wasn't initialized it would *currently at least* work ok with language server. It is initialized I think in init, so it should be covered.

On having to add the type, there was also the benefit of having a map to do lookup as opposed to a whole bunch of cases, and the implementation caches the results of RttiUtil which is perhaps a *small* benefit.

File: source/compiler-core/slang-source-map.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (jsmall-nvidia) [reply source/compiler-core/slang-source-map.cpp:214]: Copy paste I'm afraid.
